pep-0567.rst

```diff
@@ -348,48 +343,59 @@ using ``dict.copy()``::
             del copy._mapping[key]
             return copy
 
-Every OS thread has a reference to the current ``_ContextData``.
-``PyThreadState`` is updated with a new ``context_data`` field that
-points to a ``_ContextData`` object::
+Every OS thread has a reference to the current ``Context`` object::
 
     class PyThreadState:
-        context_data: _ContextData
-
-``contextvars.copy_context()`` is implemented as follows::
-
-    def copy_context():
-        ts : PyThreadState = PyThreadState_Get()
-
-        if ts.context_data is None:
-            ts.context_data = _ContextData()
-
-        ctx = Context()
-        ctx._data = ts.context_data
-        return ctx
+        context: Context
 
 ``contextvars.Context`` is a wrapper around ``_ContextData``::
 
     class Context(collections.abc.Mapping):
 
+        _data: _ContextData
+        _prev_context: Optional[Context]
+
         def __init__(self):
             self._data = _ContextData()
+            self._prev_context = None
 
         def run(self, callable, *args, **kwargs):
-            ts : PyThreadState = PyThreadState_Get()
-            saved_data : _ContextData = ts.context_data
+            if self._prev_context is not None:
+                raise RuntimeError(
+                    f'cannot enter context: {self} is already entered')
 
+            ts: PyThreadState = PyThreadState_Get()
+            if ts.context is None:
+                ts.context = Context()
+
+            self._prev_context = ts.context
             try:
-                ts.context_data = self._data
+                ts.context = self
                 return callable(*args, **kwargs)
             finally:
-                self._data = ts.context_data
-                ts.context_data = saved_data
+                ts.context = self._prev_context
+                self._prev_context = None
+
+        def copy(self):
+            new = Context()
+            new._data = self._data
+            return new
 
         # Mapping API methods are implemented by delegating
-        # `get()` and other Mapping calls to `self._data`.
+        # `get()` and other Mapping methods to `self._data`.
```

Comment:
I'd like to see these written out -- tedious and boring, but it could help getting everyone on the same page, and would avoid later surprises.

Reply:
Done: I've implemented all abstract methods in `collections.abc.Mapping`. `Context` actually subclasses it, so the implementation was indeed incomplete.